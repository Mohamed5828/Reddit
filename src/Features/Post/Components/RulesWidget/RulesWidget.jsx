import { useState } from "react";
import { RiArrowDownSLine, RiArrowUpSLine } from "react-icons/ri";
import WidgetContainer from "../WidgetContainer/WidgetContainer";

import {
  Arrow,
  Description,
  Number,
  RuleContainer,
  RuleFex,
  Title,
} from "./RulesWidget.styled";

/**
 * show rules of the community
 *
 * @returns {React.Component} CreatePost component
 */
const RulesWidget = () => {
  //rules
  const rules = [
    { title: "t1", description: "d1" },
    { title: "t2", description: "d2" },
    { title: "t3", description: "" },
  ];

  /**
   *
   * @param {number} index - order of rule
   * @param {number} len - the number of rules
   * @param {string} title - title of rule
   * @param {string} description - description of rule
   * @returns {React.Component} CreatePost component
   */
  const Rule = ({ index, len, title, description }) => {
    //to toggle the dropdown of description
    const [toggle, setToggle] = useState(true);
    //to check if will put description or not
    const desc = description.length === 0 ? false : true;
    //check the last rule so not adding line after it
    const last = index === len ? true : false;

    /**
     * onClick rule handler
     */
    function toggleHandler() {
      setToggle((prev) => !prev);
    }

    return (
      <RuleContainer onClick={toggleHandler} last={last} desc={desc}>
        <RuleFex>
          <Number>{index + "."}</Number>
          <Title>{title}</Title>
          <Arrow>
            {desc && toggle && (
              <span className="icon">
                <RiArrowDownSLine />
              </span>
            )}
            {desc && !toggle && (
              <span className="icon">
                <RiArrowUpSLine />
              </span>
            )}
          </Arrow>
        </RuleFex>
        {desc && !toggle && (
          <Description>
            <div>
              <p>{description}</p>
            </div>
          </Description>
        )}
      </RuleContainer>
    );
  };

  return (
    <WidgetContainer headerText="r/Eln2aa4yn Rules">
      {rules.map((rule, i) => {
        return (
          <Rule
            index={i + 1}
            len={rules.length}
            title={rule.title}
            key={rule.title}
            description={rule.description}
          />
        );
      })}
    </WidgetContainer>
  );
};

export default RulesWidget;
