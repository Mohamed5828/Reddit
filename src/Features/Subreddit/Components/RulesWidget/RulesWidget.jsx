import { useSubReddit } from "Features/Subreddit/Contexts/SubRedditProvider";
import { useState, useEffect } from "react";
import { RiArrowDownSLine, RiArrowUpSLine } from "react-icons/ri";
import WidgetContainer from "../WidgetContainer/WidgetContainer";

import {
  Arrow,
  Description,
  Number,
  RuleContainer,
  RuleFex,
  Title,
} from "./RulesWidget.styled";

/**
 * show rules of the community
 *
 * @returns {React.Component}
 */
const RulesWidget = () => {
  const { community } = useSubReddit();
  //rules
  // const rules = [
  //   { title: "t1", description: "d1" },
  //   { title: "t2", description: "d2" },
  //   { title: "t3", description: "" },
  // ];

  // const rules = community[0].communityRules;
  const [rules, setRules] = useState([]);
  useEffect(() => {
    community &&
      community.length &&
      setRules(community[0].communityRules || []);
  }, [community]);

  /**
   *
   * @param {number} index - order of rule
   * @param {number} len - the number of rules
   * @param {string} title - title of rule
   * @param {string} description - description of rule
   * @returns {React.Component} Rule component
   */
  const Rule = ({ index, len, title, description }) => {
    //to toggle the dropdown of description
    const [toggle, setToggle] = useState(true);
    //to check if will put description or not
    const desc = description.length === 0 ? false : true;
    //check the last rule so not adding line after it
    const last = index === len ? true : false;

    /**
     * onClick rule handler
     */
    function toggleHandler() {
      setToggle((prev) => !prev);
    }

    return (
      <RuleContainer onClick={toggleHandler} last={last} desc={desc}>
        <RuleFex>
          <Number>{index + "."}</Number>
          <Title>{title}</Title>
          <Arrow>
            {desc && toggle && (
              <span className="icon">
                <RiArrowDownSLine />
              </span>
            )}
            {desc && !toggle && (
              <span className="icon">
                <RiArrowUpSLine />
              </span>
            )}
          </Arrow>
        </RuleFex>
        {desc && !toggle && (
          <Description>
            <div>
              <p>{description}</p>
            </div>
          </Description>
        )}
      </RuleContainer>
    );
  };

  return (
    <WidgetContainer headerText="r/Eln2aa4yn Rules">
      {rules.map((rule, i) => {
        return (
          community &&
          community.length && (
            <Rule
              key={i}
              index={i + 1}
              len={rules.length}
              title={rule.title}
              description={rule.textDescription}
            />
          )
        );
      })}
    </WidgetContainer>
  );
};

export default RulesWidget;
