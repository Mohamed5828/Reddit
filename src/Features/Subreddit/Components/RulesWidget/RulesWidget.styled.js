import styled from "styled-components";

export const RuleContainer = styled.div`
  font-size: 14px;
  font-weight: 500;
  line-height: 18px;
  ${(props) =>
    props.last ? "" : "border-bottom: 1px solid rgba(26, 26, 27, 0.07);"}
  ${(props) => (props.last ? "padding: 8px 0 0 0;" : "padding: 8px 0;")}
  cursor: ${(props) => (props.desc ? "pointer" : "default")};
`;

export const RuleFex = styled.div`
  display: flex;
  flex-direction: row;
`;

export const Number = styled.div`
  box-sizing: border-box;
  flex: 0 0;
  padding-right: 2px;
`;

export const Title = styled.div`
  box-sizing: border-box;
  flex: 1 1 100%;
  width: 100%;
  padding-right: 8px;
  word-break: break-word;
`;

export const Arrow = styled.div`
  flex: 0 0;
  span.icon {
    font-size: 20px;
    font-weight: 400;
    height: 20px;
    line-height: 20px;
    vertical-align: middle;
    width: 20px;
  }
`;

export const Description = styled.div`
  font-family: "Noto Sans", Arial, sans-serif;
  font-size: 12px;
  font-weight: 400;
  line-height: 16px;
  padding: 8px 8px 0 16px;

  div {
    line-height: 21px;
  }

  p {
    font-size: 14px;
    line-height: 21px;
    word-break: break-word;
    padding-top: 0;
    padding-bottom: 0;
    margin-bottom: 0;
  }
`;
