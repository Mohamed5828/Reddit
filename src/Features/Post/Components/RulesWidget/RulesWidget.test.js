import Enzyme, { shallow } from "enzyme";
import RulesWidget from "./RulesWidget";

describe("widget show rules", () => {
  it("should render without crashing", () => {
    expect(shallow(<RulesWidget />)).toMatchSnapshot();
  });
});
