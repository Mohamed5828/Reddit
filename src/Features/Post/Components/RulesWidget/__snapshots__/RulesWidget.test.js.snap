// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`widget show rules should render without crashing 1`] = `
<WidgetContainer
  headerText="r/Eln2aa4yn Rules"
>
  <Rule
    description="d1"
    index={1}
    key="t1"
    len={3}
    title="t1"
  />
  <Rule
    description="d2"
    index={2}
    key="t2"
    len={3}
    title="t2"
  />
  <Rule
    description=""
    index={3}
    key="t3"
    len={3}
    title="t3"
  />
</WidgetContainer>
`;
